### ISSUE TITLE
The obfuscar builded from the github source code can NOT work well

### ISSUE THREAD

**Issue by cqgreen** [lextm/obfuscar#69]
I build obfuscar with the source from github. It can NOT work well for [my dll attached](http://share.weiyun.com/30ae1a46c3e47368e1ce7a7951b0f032).
The ILSpy show the following error, when I check the class SoldierAttribute in Assembly-CSharp.dll.
But the 2.1.1 release from CodePlex workes well.

`ICSharpCode.Decompiler.DecompilerException: Error decompiling System.Void c.l::A(C.J,System.Boolean)
 ---> System.ArgumentNullException: 值不能为 null。
参数名: key
   在 System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   在 System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   在 ICSharpCode.Decompiler.ILAst.SimpleControlFlow.SimplifyShortCircuit(List`1 body, ILBasicBlock head, Int32 pos)
   在 ICSharpCode.Decompiler.ILAst.ILAstOptimizerExtensionMethods.RunOptimization(ILBlock block, Func`4 optimization)
   在 ICSharpCode.Decompiler.ILAst.ILAstOptimizer.Optimize(DecompilerContext context, ILBlock method, ILAstOptimizationStep abortBeforeStep)
   在 ICSharpCode.Decompiler.Ast.AstMethodBodyBuilder.CreateMethodBody(IEnumerable`1 parameters)
   在 ICSharpCode.Decompiler.Ast.AstMethodBodyBuilder.CreateMethodBody(MethodDefinition methodDef, DecompilerContext context, IEnumerable`1 parameters)
   --- 内部异常堆栈跟踪的结尾 ---
   在 ICSharpCode.Decompiler.Ast.AstMethodBodyBuilder.CreateMethodBody(MethodDefinition methodDef, DecompilerContext context, IEnumerable`1 parameters)
   在 ICSharpCode.Decompiler.Ast.AstBuilder.CreateMethod(MethodDefinition methodDef)
   在 ICSharpCode.Decompiler.Ast.AstBuilder.AddTypeMembers(TypeDeclaration astType, TypeDefinition typeDef)
   在 ICSharpCode.Decompiler.Ast.AstBuilder.CreateType(TypeDefinition typeDef)
   在 ICSharpCode.Decompiler.Ast.AstBuilder.AddType(TypeDefinition typeDef)
   在 ICSharpCode.ILSpy.CSharpLanguage.DecompileType(TypeDefinition type, ITextOutput output, DecompilationOptions options)
   在 ICSharpCode.ILSpy.TextView.DecompilerTextView.DecompileNodes(DecompilationContext context, ITextOutput textOutput)
   在 ICSharpCode.ILSpy.TextView.DecompilerTextView.<>c__DisplayClass16.<DecompileAsync>b__15()
`

**Comment by lextm** [comment]
Thanks. Just get some time and will check it out.

**Comment by lextm** [comment]
I just played with the test assemblies. It seems to me that at least dotPeek can show the proper IL contained in the obfuscated method, though both dotPeek and ILSpy cannot decompile it back to C#.

Then the question to you @cqgreen is whether CLR thinks the obfuscated IL is valid or not. When you run the obfuscated application, is there any exception related to this method?

If it is valid, then I don't think it is a must for ILSpy or dotPeek to decompile. If it is invalid, then we might further investigate.

Meanwhile, I tested 2.2.0 and 2.1.x releases, and the obfuscated results from both releases trigger such an exception in ILSpy. It seems that only 2.0.3 release can obfuscate in a way that ILSpy can decompile, which I will investigate further too.
